hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/om/FinalizeUpgradeSubCommand.java

```diff
@@ -0,0 +1,221 @@
+ * the License.
+ */
+
+package org.apache.hadoop.ozone.admin.om;
+
+import org.apache.hadoop.hdds.cli.HddsVersionProvider;
+import org.apache.hadoop.ozone.om.exceptions.OMException;
+import org.apache.hadoop.ozone.om.protocol.OzoneManagerProtocol;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.UpgradeFinalizationStatus;
+import picocli.CommandLine;
+
+import java.io.IOException;
+import java.util.UUID;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+import static org.apache.hadoop.ozone.om.exceptions.OMException.ResultCodes.INVALID_REQUEST;
+
+/**
+ * Handler of ozone admin om finalizeUpgrade command.
+ */
+@CommandLine.Command(
+    name = "finalizeupgrade",
+    description = "Finalizes Ozone Manager's metadata changes and enables new "
+        + "features after a software upgrade.\n"
+        + "It is possible to specify the service ID for an HA environment, "
+        + "or the Ozone manager host in a non-HA environment, if none provided "
+        + "the default from configuration is being used if not ambiguous.",
+    mixinStandardHelpOptions = true,
+    versionProvider = HddsVersionProvider.class
+)
+public class FinalizeUpgradeSubCommand implements Callable<Void> {
+
+  @CommandLine.ParentCommand
+  private OMAdmin parent;
+
+  @CommandLine.Option(
+      names = {"-id", "--service-id"},
+      description = "Ozone Manager Service ID"
+  )
+  private String omServiceId;
+
+  @CommandLine.Option(
+      names = {"-host", "--service-host"},
+      description = "Ozone Manager Host"
+  )
+  private String omHost;
+
+  @CommandLine.Option(
+      names = {"--takeover"},
+      description = "Forces takeover of monitoring from an other client, if "
+          + "finalization has already been started and did not finished yet."
+  )
+  private boolean force;
+
+  @Override
+  public Void call() throws Exception {
+    boolean forceHA = false;
+    OzoneManagerProtocol client =
+        parent.createOmClient(omServiceId, omHost, forceHA);
+    String upgradeClientID = "Upgrade-Client-" + UUID.randomUUID().toString();
+    try {
+      UpgradeFinalizationStatus status =
+          client.finalizeUpgrade(upgradeClientID);
+      if (isFinalized(status)){
+        System.out.println("Upgrade has already been finalized.");
+        emitExitMsg();
+        return null;
+      } else if (!isStarting(status)){
+        System.err.println("Invalid response from Ozone Manager.");
+        System.err.println(
+            "Current finalization status is: " + status.getStatus()
+        );
+        throw new IOException("Exiting...");
+      }
+    } catch (OMException e) {
+      handleInvalidRequestAfterInitiatingFinalization(e);
+    }
+    monitorAndWaitFinalization(client, upgradeClientID);
```

Comment:
Sometimes users just want to query immediate upgrade status and don't want to wait and monitor finalization finished. Can we add one additionally option that we can skip monitorAndWaitFinalization?

Reply:
There is a plan to expose the current status via JMX(HDDS-4142), would that be a sufficient way to provide this functionality?
If JMX is not enough to expose a queryable state, this might come in as an improvement for the client side. I am happy to implement it here but I would get back to this after the core logic is working in a separate JIRA if you don't mind, let me know if that works for you, and I will file the JIRA.

Reply:
Yes, of course, we can address above suggestion later. That's just an improvement.